### ISSUE TITLE
High memory usage of CSV sorting

### ISSUE THREAD

**Issue by minimum-necessary-change** [BurntSushi/imdb-rename#19]
Hello,

I was trying to use the application on a Linkstation LS-QVL(ARMv5, 256MB ram) running Debian Buster.

`
$ uname -a
Linux nasprimary 4.19.0-14-marvell # 1 Debian 4.19.171-2 (2021-01-30) armv5tel GNU/Linux

$ cat /proc/cpuinfo 
processor	: 0
model name	: Feroceon 88FR131 rev 1 (v5l)
BogoMIPS	: 400.00
Features	: swp half thumb fastmult edsp 
CPU implementer	: 0x56
CPU architecture: 5TE
CPU variant	: 0x2
CPU part	: 0x131
CPU revision	: 1

Hardware	: Marvell Kirkwood (Flattened Device Tree)
Revision	: 0000
Serial		: 0000000000000000

$ free
              total        used        free      shared  buff/cache   available
Mem:         254040       82320       35056         964      136664      162848
Swap:       1000432        3592      996840
`
 
While it builds natively and runs. It quickly consumes all available memory sorting the downloaded CSV file.

Ref: https://github.com/BurntSushi/imdb-rename/issues/12

Due to the high memory requirements of trying to load the entire CSV file into memory, would it not be possible to leave the file on disk and build an index in memory? It would be slower but would have to be done only once?

Building the index in memory while leaving the huge CSV file on disk would also the application to work of small SBC computers such as the Raspberry Pi, Orange Pi and others which typically ship with 256MB-2GB ram (ignoring the 4GB RPi).

Thanks

**Comment by BurntSushi** [comment]
The high memory usage was intentional for implementation simplicity. If you can come up with a small and simple patch that fixes your problem, then I'd be okay with that. It's been a while since I've touched this code, but I don't think such a patch exists. Otherwise, I have no intention of fixing this. Sorry.

**Comment by minimum-necessary-change** [comment]
Ok, understandable.

I'll leave the below if anybody comes across this issue in future

How to index a csv file:
https://stackoverflow.com/questions/36089456/indexing-columns-in-a-csv-file/46253244

**Comment by BurntSushi** [comment]
@minimum-necessary-change "creating an index file" is not a solution to this problem. In fact, that's basically the entire point of `imdb-rename`. Most of the code in this project is specifically for _creating an index_ into the CSV files. How else would querying be instantaneous? The process of creating an index is actually already optimized to not require the entire corpus to be in memory at once. Unfortunately, in order to do that, it requires that the CSV records be in lexicographically sorted order. This _used_ to be the case, but the source of the data changed such that it was no longer sorted. I then implemented a simple fix in commit 09e07a22, which was to load the entire data in memory and sort it. That initial implementation used more memory than necessary for this specific corpus though, and commit 2ea88a08 fixed that.

So "indexing a csv file" is missing the point here. The actual issue influencing memory usage is sorting the data. Fixing that requires doing sorting of data with a fixed memory usage limit. Such a thing is of course possible. But it is much more complex than loading everything into memory. In principle, tools like GNU sort do this, but I am not interested in shelling out to such tools in this program.

To summarize, if one can create a simple patch that sorts the lines of a file with bounded memory usage without sacrificing too much performance, then I'd accept that. If the simple patch requires the use of a third party crate, then that crate needs to pass my quality standards.

**Comment by minimum-necessary-change** [comment]
Thanks for the info.

As a test I imported the TSV into sqlite3 on the low memory NAS and selected "Battlefield Earth" as a quick test. The results are below. Although the import wasn't in anyway quick the memory impact was minimal.

I really am not a script programmer so my Rust is pretty sparse. I can read it but my API knowledge is lacking.

`$ sqlite3
sqlite> create table title (titleId text, ordering int, title text, region text, language text, types text, attributes text, isOriginalTitle int);
sqlite> .separator "\t"
sqlite> .import title.akas.tsv title
sqlite> select title,ordering from title where title = "Battlefield Earth" order by ordering asc;
Battlefield Earth	1
Battlefield Earth	5
Battlefield Earth	12
Battlefield Earth	16
Battlefield Earth	18
Battlefield Earth	22
Battlefield Earth	32
sqlite> .exit

$ ls -al ./
total 2587972
drwxr-xr-x  2 mnc mnc           41 Mar  3 14:25 .
drwxr-xr-x 11 mnc users        148 Mar  3 14:05 ..
-rw-r--r--  1 mnc mnc   1403027456 Mar  3 14:25 imdb.db
-rw-r--r--  1 mnc mnc   1247057560 Feb 14 13:20 title.akas.tsv

$ free
              total        used        free      shared  buff/cache   available
Mem:         254040       74764       16412        1044      162864      169912
Swap:       1000432        7564      992868
`
Thanks.

**Comment by BurntSushi** [comment]
I don't understand what you're trying to show. You're using sqlite3 as an elaborate implementation of out-of-core sorting, which is exactly what I said was needed here. (Short of a more clever approach that utilizes properties of the source data, if they exist.)

**Comment by minimum-necessary-change** [comment]
I apologise in that we seem to be slightly at odds. I am simply looking for a solution to a memory usage problem.

I started out programming the Atari 400, Commodore 128, Amigas and Atari ST, using tape drives, floppy disks and my first hard drive was a mere 10MB. Memory was a premium and so when I see a CSV/TSV file requiring 1.7GB to be sorted when I believe all that was required is to obtain a sorted index onto that CSV fixed data. I find myself perplexed.

By using the command-line interface, I was trying to establish whether it was feasible to use the tried and tested sqlite database library to effectively sort and index the IMDB data in an efficient manner on a low memory machine.

I have also just now found a rust (module?) that handles sqlite3 databases @ https://rust-lang-nursery.github.io/rust-cookbook/database/sqlite.html

None of this information I knew earlier this morning!

`
$ cat import_tsv.sh
#!/bin/bash

sqlite3 imdb2.db -cmd "create table title (titleId text, ordering int, title text, region text, language text, types text, attributes text, isOriginalTitle int);" ".exit"
sqlite3 imdb2.db -cmd ".separator \"\t\"" ".import title.akas.tsv title" ".exit"
sqlite3 imdb2.db -cmd "select title,ordering from title where title = 'Battlefield Earth' order by ordering asc;" ".exit"

$ time ./import_tsv.sh
title.akas.tsv:155487: unescaped " character
title.akas.tsv:155490: unescaped " character
title.akas.tsv:155490: unescaped " character
. . .
. . .
title.akas.tsv:24974908: unescaped " character
title.akas.tsv:25062776: unescaped " character
title.akas.tsv:25064331: unescaped " character
Battlefield Earth|1
Battlefield Earth|5
Battlefield Earth|12
Battlefield Earth|16
Battlefield Earth|18
Battlefield Earth|22
Battlefield Earth|32

real	15m46.520s
user	15m2.422s
sys	0m29.171s

$ time sqlite3 imdb2.db -cmd "select title,ordering from title where title = 'Battlefield Earth' order by ordering asc;" ".exit"
Battlefield Earth|1
Battlefield Earth|5
Battlefield Earth|12
Battlefield Earth|16
Battlefield Earth|18
Battlefield Earth|22
Battlefield Earth|32

real	0m35.261s
user	0m24.971s
sys	0m9.199s
`

Thanks

**Comment by BurntSushi** [comment]
Yes, we're at odds. Writing code for memory constrained systems when dealing with corpora that don't fit into the memory of said systems is harder than writing code for systems that can fit said corpora into memory. You pointing at sqlite3 and saying, "see it's possible" is actually _supporting_ that point. sqlite3 is a monstrously complicated embedded database. The entire _point_ of sqlite3 is to be out-of-core!

In case you're saying, "well, then please just use sqlite3." Well, no, sqlite3 doesn't support the kind of information retrieval techniques used by imdb-rename. Moreover, sqlite3 is not a dependency I want to add to this tool. Using sqlite3 would imply significant design changes to how this tool works.

Upon reviewing the code, I was mistaken earlier when I said that the rest of the index creation process was designed such that it doesn't require putting some significant fraction of the corpus in memory. Thus, fixing just the CSV sorting won't resolve this problem as far as I can tell. It looks like there are other portions of the index creation process that do additional sorting, so those would in turn also need to be converted to out-of-core sorts.

I see no simple patch to fix your problem.

**Comment by minimum-necessary-change** [comment]
Ok, thanks for taking a deeper look at this.
We can probably let it go now.
